File: section1/6charmin9-dev/_1st_week/_01_06_find_prime_list_under_number.py
input = 20


def find_prime_list_under_number(number):
    prime_list = []
    for num in range(2, number + 1): #N
        for prime in prime_list: #N [2, 3, 5, 7, 11, 13, 17, 19]
            if prime * prime <= num and num % prime == 0: # 소수가 아닌 조건
                break
        else:
            prime_list.append(num)
    return prime_list



result = find_prime_list_under_number(input)
print(result)

# 시간복잡도: O(N^2) -> O(Nroot(N))
# 공간복잡도: O(N)

File: section1/6charmin9-dev/_1st_week_java/_01_06_find_prime_list_under_number.java
/*
Q. 정수를 입력 했을 때, 그 정수 이하의 소수를 모두 반환하시오. 

소수는 자신보다 작은 두 개의 자연수를 곱하여 만들 수 없는 1보다 큰 자연수이다.
# 20이 입력된다면, 아래와 같이 반환해야 합니다!
[2, 3, 5, 7, 11, 13, 17, 19]
 */

import java.util.ArrayList;
import java.util.List;

public class _01_06_find_prime_list_under_number extends TempleteUtils {
    public static void main(String[] args) {
        int input = 20;

        print(find_prime_list_under_number(input));
    }

    public static int[] find_prime_list_under_number(int number) {
        List<Integer> primes = new ArrayList<>();
        for (int i=2; i<=number; i++) {
            boolean isPrime = true;
            for (int prime : primes) {
                if (prime * prime <= i && i % prime == 0) {
                    isPrime = false;
                    break;
                }
            }
            if (isPrime) {
                primes.add(i);
            }
        }
        return primes.stream().mapToInt(i -> i).toArray();
    }
}


